src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -1798,6 +1798,20 @@ cdef class fmpz_mod_poly(flint_poly):
         if not self.ctx.is_prime():
             raise NotImplementedError("factor algorithm assumes that the modulus is prime")
```

Comment:
The new `fmpz_mod_mpoly` and `nmod_mpoly` are inconsistent with this exception type, they raise a `DomainError` instead.  Seems like this isn't consistent across the code base, we have `DomainError`, `NotImplementedError`, and `ValueError` for similar exception reasons. What should the preference here?

```
Exported grep results:

src/flint/types/nmod_mpoly.pyx:312:            raise DomainError("division with non-prime modulus is not supported")
src/flint/types/nmod_mpoly.pyx:759:            raise DomainError("gcd with non-prime modulus is not supported")
src/flint/types/nmod_mpoly.pyx:780:            raise DomainError("square root with non-prime modulus is not supported")
src/flint/types/nmod_mpoly.pyx:811:            raise DomainError("factorisation with non-prime modulus is not supported")
src/flint/types/nmod_mpoly.pyx:853:            raise DomainError("factorisation with non-prime modulus is not supported")
src/flint/types/fmpz_mod.pyx:344:            raise NotImplementedError("algorithm assumes modulus is prime")
src/flint/types/fq_default.pyx:98:                    raise ValueError("characteristic is not prime")
src/flint/types/fq_default.pyx:108:            raise ValueError("either a prime or modulus must be passed for construction")
src/flint/types/fq_default.pyx:133:            raise ValueError("characteristic is not prime")
src/flint/types/fq_default.pyx:156:            raise ValueError("characteristic is not prime")
src/flint/types/fmpz_mod_poly.pyx:285:            raise NotImplementedError("minpoly algorithm assumes that the modulus is prime")
src/flint/types/fmpz_mod_poly.pyx:1270:            raise NotImplementedError("gcd algorithm assumes that the modulus is prime")
src/flint/types/fmpz_mod_poly.pyx:1765:            raise NotImplementedError("factor_squarefree algorithm assumes that the modulus is prime")
src/flint/types/fmpz_mod_poly.pyx:1799:            raise NotImplementedError("factor algorithm assumes that the modulus is prime")
src/flint/types/fmpz_mod_poly.pyx:1844:            raise NotImplementedError("factor algorithm assumes that the modulus is prime")
src/flint/types/fmpz_mod_mpoly.pyx:333:            raise DomainError("division with non-prime modulus is not supported")
src/flint/types/fmpz_mod_mpoly.pyx:783:            raise DomainError("gcd with non-prime modulus is not supported")
src/flint/types/fmpz_mod_mpoly.pyx:804:            raise DomainError("square root with non-prime modulus is not supported")
src/flint/types/fmpz_mod_mpoly.pyx:835:            raise DomainError("factorisation with non-prime modulus is not supported")
src/flint/types/fmpz_mod_mpoly.pyx:878:            raise DomainError("factorisation with non-prime modulus is not supported")
src/flint/test/test_all.py:3757:    assert raises(lambda: flint.fq_default_ctx(modulus=mod_not_prime), ValueError)
```

Reply:
I think that there should be a special exception type like `NonPrimeModulusError` because it is such a common case in the codebase.

It is also something that much real downstream code will never had to deal with because using a non-prime modulus is not that common or at least if using it then you would already know know not to use things like division.